test/bio_read_test.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <stdio.h>
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/bio.h>
+#include <openssl/rand.h>
+#include <openssl/bio.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include "testutil/output.h"
+#include "testutil.h"
+
+#define PACKET_COUNT 10
+
+/*
+ * this test case opens a pair of v4 or v6 sockets, bound to localhost.
+ * A process is forked to do the writing. The write path is tested in
+ * bio_write_test case.
+ *
+ * This code exercises the BIO_read side of things.
+ *
+ * The foreground process reads 5 packets from the socket and discards
+ * them.
+ *
+ * The packets being send to the other destination are just ignored for now.
+ */
+
+static int read_socket_and_discard(int fd, int count, unsigned portnum)
+{
+    char buf[512];
+    BIO *in;
+    int ret;
+    BIO_ADDR *ba;
+
+    /* do not wait forever! */
+    alarm(60);
+
+    ba = BIO_ADDR_new();
+
+    in = BIO_new_dgram(fd, BIO_CLOSE);
+
+    while (--count > 0 && (ret = BIO_read(in, buf, 512)) > 0) {
+
+        /* now check out the bio structure for the origin of the packet */
+        BIO_get_dgram_origin(in, ba);
+        unsigned short port = ntohs(BIO_ADDR_rawport(ba));
+        if (port != 0 && port != portnum) {
+            exit(5);
+        }
+    }
+
+    if (ret <= 0) {
+        test_printf_stderr("BIO_read returned %d\n", ret);
+    }
+
+    BIO_ADDR_free(ba);
+
+    return count;
+}
+
+static int write_packets(BIO *bio, int count, BIO_ADDR *dst1, BIO_ADDR *dst2)
+{
+    const char hello[] = "hellohellohello";
+    int toggle = 0;
+
+    while (--count > 0) {
+        const int sizeofhello = sizeof(hello);
+
+        if (toggle && dst2 != NULL) {
+            BIO_set_dgram_dest(bio, dst2);
+        } else {
+            BIO_set_dgram_dest(bio, dst1);
+        }
+
+        int ret = BIO_write(bio, hello, sizeofhello);
+        if (ret != sizeofhello) {
+            exit(3);
+        }
+
+        toggle = !toggle;
+    }
+
+    return 0;
+}
+
+static int fork_and_read_write_packets(int infd, int outfd,
+                                       unsigned portnum,
+                                       BIO_ADDR *dsthost1, BIO_ADDR *dsthost2)
+{
+    BIO *out;
+    int expected_count = PACKET_COUNT;
+
+    pid_t pid = fork();
```

Comment:
fork() doesn't exist on Windows!

Reply:
@mattcaswell, my run of the tests on travis seemed to just die in the middle without any obvious error.  I guess AppVeyor builds for Windows?  
I also now recall that travis has no IPv6. Not sure what to do about that yet.  
I don't want to rewrite this test case with select(), because that will just be more complicated and won't work on Windows anyway.  Also, re, //- comments, shouldn't the default flags warn on that if travis builds break on that?
Finally, 90% of the style issues you complain about were introduced by openssl-format-source. So I think that openssl-format-source does not comply to the style guide.
What I did was to run it without any material changes and then commit those as on their own so that my changes would be not obscured by indent changes.  Why is it wrong to fix bss_dgram.c and bio.h's indented?

Reply:
> I guess AppVeyor builds for Windows?

Yes

> I also now recall that travis has no IPv6. Not sure what to do about that yet.

Possibly if we can't create the sockets then we skip that bit of the test - but don't fail it.

Reply:
> Also, re, //- comments, shouldn't the default flags warn on that if travis builds break on that?

The default flags are designed for the general user - not people developing patches. OpenSSL has to run on lots of different platforms with lots of different compilers. We don't run the default with too high warning options because that will introduce spurious build failures for the general user. If you are developing patches you should use `--strict-warnings` as mentioned in the CONTRIBUTING file.

> Finally, 90% of the style issues you complain about were introduced by openssl-format-source. So I think that openssl-format-source does not comply to the style guide.

Probably it doesn't. The style has developed further since that was written. I wouldn't recommend using it myself. As I said above I would back out those changes.

> Why is it wrong to fix bss_dgram.c and bio.h's indented?

It generates a lot of noise and makes it difficult to review the patch (difficult to distinguish your changes from auto-generated changes). If we want to fix large amounts of the formatting in files it should be done in a dedicated PR. Localised fixes are ok.

Reply:
I have removed the patch that just ran openssl-format-source on that code.